src/async_fuse/memfs/dist/lock_manager.rs

```diff
@@ -0,0 +1,384 @@
+            })?;
+
+        // cancel signal
+        let (cancel_tx, cancel_rx) = sync::oneshot::channel::<()>();
+
+        self.locked
+            .write()
+            .timeout_renews
+            .insert(key_vec, (cancel_tx, lockinfo.unlock_token));
+        self.register_renew_tx
+            .send(LockDescription {
+                lease_keeper,
+                renew_timeout: DEFAULT_RENEW_TIMEOUT,
+                cancel_rx,
+            })
+            .await
+            .unwrap_or_else(|e| {
+                panic!("lock manager: register renew rx is dropped, which is impossible, err:{e}")
+            });
+        Ok(true)
+    }
+    /// Unlock dist lock
+    /// - if locked before return true
+    /// - if not locked before return false
+    pub async fn unlock(&self, key: &LockKeyType) -> DatenLordResult<bool> {
+        log::debug!("`LockManager::unlock`, key: {key:?}");
+        let key: Vec<u8> = key.get_key();
+        let take_lock = self.locked.write().timeout_renews.remove_entry(&key);
+        if let Some((key, (cancel_sender, unlock_token))) = take_lock {
+            cancel_sender
+                .send(())
+                .unwrap_or_else(|_| panic!("lock manager cancel receiver dropped, which should be there until this function is called"));
+            self.kv_engine
+                .unlock(unlock_token)
+                .await
+                .with_context(|| format!("lock manager: etcd unlock failed, lock key: {key:?}"))?;
+            Ok(true)
+        } else {
+            Ok(false)
+        }
+    }
+    /// Stop the async task
+    pub async fn system_end(&self) {
+        let tx =
+            self.locked.write().system_end_tx.take().unwrap_or_else(|| {
+                panic!("lock manager: system end tx is taken, which is impossible")
+            });
+        tx.send(()).unwrap_or_else(|_| {
+            panic!("lock manager: system end rx is dropped, which is impossible")
+        });
+    }
+}
+
+#[cfg(test)]
+#[allow(clippy::unwrap_used)]
+mod test {
+    use std::sync::Arc;
+
+    use clippy_utilities::OverflowArithmetic;
+
+    use super::*;
+    use crate::{
+        async_fuse::{
+            fuse::file_system::new_fs_async_result_chan,
+            memfs::kv_engine::{EtcdKVEngine, LockKeyType},
+        },
+        common,
+    };
+
+    const ETCD_ADDRESS: &str = "localhost:2379";
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn test_lockmanager() {
+        common::util::setup_test_log_debug();
+        for i in 0_i32..6_i32 {
+            log::debug!("test_lockmanager_once for the {} time", i.overflow_add(1));
+            test_lockmanager_once().await;
+        }
+    }
+
+    // #[tokio::test(flavor = "multi_thread")]
+    #[allow(clippy::unwrap_used, clippy::too_many_lines)]
+    async fn test_lockmanager_once() {
+        common::util::setup_test_log_debug();
+
+        let mut server = mock_etcd::MockEtcdServer::default();
```

Comment:
We can move these preparation tasks to another function, which may be reused in the future. This also keeps the testing logic clean.

Reply:
Agree with that.